src/createWidgetBase.ts

```diff
@@ -221,40 +261,53 @@ const createWidget: WidgetFactory = createStateful
 					return assign(baseIdProp, { key: this, classes, styles });
 				}
 
-			],
+		],
 
-			__render__(this: Widget<WidgetState>): VNode | string | null {
+			__render__(this: Widget<WidgetState, WidgetProperties>): VNode | string | null {
 				const internalState = widgetInternalStateMap.get(this);
+				const updatedProperties = generateProperties(this, internalState.previousProperties);
+				this.applyChangedProperties(updatedProperties.previousProperties, updatedProperties.currentProperties);
+
 				if (internalState.dirty || !internalState.cachedVNode) {
 					const widget = dNodeToVNode(this, this.getNode());
 					manageDetachedChildren(this);
 					if (widget) {
 						internalState.cachedVNode = widget;
 					}
 					internalState.dirty = false;
+					internalState.previousProperties = this.properties;
 					return widget;
 				}
 				return internalState.cachedVNode;
 			},
 
-			get registry(this: Widget<WidgetState>): FactoryRegistry {
+			get registry(this: Widget<WidgetState, WidgetProperties>): FactoryRegistry {
 				return widgetInternalStateMap.get(this).factoryRegistry;
 			},
 
 			tagName: 'div'
 		},
-		initialize(instance: Widget<WidgetState>, options: WidgetOptions<WidgetState> = {}) {
-			const { id, tagName } = options;
+		initialize(instance: Widget<WidgetState, WidgetProperties>, options: WidgetOptions<WidgetState, { id?: string }> = {}) {
+			const { tagName, properties = {} } = options;
+			const id = properties.id || options.id || generateID(instance);
+
+			if (!properties.id) {
```

Comment:
Okay think I have a slightly cleaner work around for this above (to avoid the options peeling)